Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp

```diff
@@ -419,30 +419,38 @@ void TextureMapperLayer::collectDamageSelf(TextureMapperPaintOptions& options, D
     transform.multiply(options.transform);
     transform.multiply(m_layerTransforms.combined);
 
-    ASSERT(!m_damage.isInvalid());
-    ASSERT(!m_inferredDamage.isInvalid());
-    if (!m_inferredDamage.isEmpty()) {
-        for (const auto& rect : m_inferredDamage.rects()) {
+    ASSERT(!m_receivedDamage.isInvalid());
+    ASSERT(!m_inferredLayerDamage.isInvalid());
+    ASSERT(!m_inferredGlobalDamage.isInvalid());
+    // Inferred damage always takes precedence over any other kind of damage as it stores
+    // the damage that is a result of the layer-level operations.
+    if (!m_inferredGlobalDamage.isEmpty() || !m_inferredLayerDamage.isEmpty()) {
+        for (const auto& rect : m_inferredGlobalDamage.rects()) {
             ASSERT(!rect.isEmpty());
             damage.add(rect);
         }
+        for (const auto& rect : m_inferredLayerDamage.rects()) {
+            ASSERT(!rect.isEmpty());
+            damage.add(transformRectFromLayerToGlobalCoordinateSpace(rect, transform, options));
+        }
     } else if (m_contentsLayer) {
         // Layers with content layer are fully damaged if there's no explicit damage.
         // FIXME: Remove that special case.
-        damage.add(transformRectForDamage(targetRect, transform, options));
+        damage.add(transformRectFromLayerToGlobalCoordinateSpace(targetRect, transform, options));
     } else {
         // Use the damage information we received from the GraphicsLayer
```

Comment:
```suggestion
        // Use the damage information reported by the GraphicsLayer
```

...if we rename it as suggested below. We could also go with a longer name, something like `m_graphicsLayerReportedDamage`?  then we can delete the comment (this would probably mean we would use `m_layerInferredDamage` and `m_globalInferredDamage` to be consistent)